### DISCUSSION TITLE
Alternative audio backends

### DISCUSSION THREAD

**Discussion by H-M-H** [jpochyla/psst#167] [General]
Continuation of the discussion about alternative audio backends started here: https://github.com/jpochyla/psst/issues/54#issuecomment-910178007

I have done some experiments with alternative audio backends. Specifically `cpal` + `dasp` on this branch: https://github.com/H-M-H/psst/tree/cpal and `rodio` here: https://github.com/H-M-H/psst/tree/rodio

This has been motivated by miniaudio not working on my system with pipewire, related issue can be found here: https://github.com/jpochyla/psst/issues/163

I'd love if someone could test this on Windows or macOS and report back on audio quality as well as CPU consumption.

On my Linux system the `cpal` branch takes about 3.3% CPU, `rodio` 4% CPU, both without resampling.

For some reason the channels using rodio are inverted though, perhaps related to: https://github.com/RustAudio/rodio/issues/286

**Comment by jpochyla** [comment]
Thanks a lot for the exploratory work @H-M-H! As I've mentioned, I've tried to use `rodio` in the beginning, but the CPU usage was considerably higher than what we have now, although it might have been because of `lewton` crate that it used for the Vorbis codec. In debug mode, i.e., it couldn't keep the buffers full and the playback was choppy. To me, it seems that a better idea would be to use `cpal` + `dasp` + `symphonia` for the audio stack, as that would give us the most flexibility in maneuvering around the issues. Tomorrow I will start working on this, based on your `cpal` branch.

**Comment by jacksongoode** [comment]
> Continuation of the discussion about alternative audio backends started here: [#54 (comment)](https://github.com/jpochyla/psst/issues/54#issuecomment-910178007)
> 
> I have done some experiments with alternative audio backends. Specifically `cpal` + `dasp` on this branch: https://github.com/H-M-H/psst/tree/cpal and `rodio` here: https://github.com/H-M-H/psst/tree/rodio
> 
> This has been motivated by miniaudio not working on my system with pipewire, related issue can be found here: #163
> 
> I'd love if someone could test this on Windows or macOS and report back on audio quality as well as CPU consumption.
> 
> On my Linux system the `cpal` branch takes about 3.3% CPU, `rodio` 4% CPU, both without resampling.
> 
> For some reason the channels using rodio are inverted though, perhaps related to: [RustAudio/rodio#286](https://github.com/RustAudio/rodio/issues/286)

Did you test the CPU utilization with the miniaudio as well? Just curious, as I found cpal from your build to be a touch higher.

**Comment by H-M-H** [comment]
> Did you test the CPU utilization with the miniaudio as well?

I could not at the time due to #163 but as things are working again I had a look. I have rebased `cpal` on the latest master and comparing I get typically a little under 3% CPU utilization using `cpal` and about 6% using miniaudio from `master`.

**Comment by jpochyla** [comment]
Hi, I'm sorry for being a bit unavailable lately. There's an experimental #197 ready with a `cpal` backend.

**Comment by jpochyla** [comment]
So I've merged the branch using cpal and symphonia into master. There's some resampling using `libsamplerate`, so far without configurable quality. I'd also like to add preferred audio device to the config, and investigate if we can react to device hotplug events (like pausing the playback).

That said, [cubeb-rs](https://github.com/mozilla/cubeb-rs) from Mozilla looks like it could be much better than CPAL once it matures.

Thank you everybody!
